src/workerd/jsg/jsg-test.c++

```diff
@@ -456,6 +458,47 @@ KJ_TEST("External memory adjustment") {
   });
 }
 
+KJ_TEST("External memory adjustment - defered") {
+  auto target = kj::refcounted<ExternalMemoryPtr>();
+
+  {
+    IsolateUuidIsolate isolate(v8System, kj::heap<IsolateObserver>());
+    target = isolate.runInLockScope(
+        [&](IsolateUuidIsolate::Lock& lock) { return lock.getExternalMemoryTarget(); });
+
+    // Adjustment to memory while not holding lock will be applied later
+    auto adjuster1 = target->getAdjustment(1000);
+    KJ_ASSERT(adjuster1.getAmount() == 1000);
+    KJ_ASSERT(target->getPendingMemoryUpdate() == 1000);
+
+    {
+      // This adjustment has no effect because the adjuster is destroyed before we take the lock again
+      auto adjuster2 = target->getAdjustment(1000);
+      KJ_ASSERT(adjuster2.getAmount() == 1000);
+      KJ_ASSERT(target->getPendingMemoryUpdate() == 2000);
+    }
+
+    KJ_ASSERT(target->getPendingMemoryUpdate() == 1000);
+    KJ_ASSERT(target->isIsolateAlive());
+
+    isolate.runInLockScope([&](IsolateUuidIsolate::Lock& lock) {
+      // Once lock is taken, the amount is applied
+      KJ_ASSERT(target->getPendingMemoryUpdate() == 0);
+
+      // Adjustment made while holding lock applies immediately
+      adjuster1.adjust(-500);
+      KJ_ASSERT(adjuster1.getAmount() == 500);
+      KJ_ASSERT(target->getPendingMemoryUpdate() == 0);
+      KJ_ASSERT(target->isIsolateAlive());
+    });
+  }
+
+  KJ_ASSERT(!target->isIsolateAlive());
+  // Making an adjustment anyway won't do anything but also won't crash
+  auto adjuster3 = target->getAdjustment(500);
+  KJ_ASSERT(target->getPendingMemoryUpdate() == 0);
+}
```

Comment:
As suggested in chat... let's add a case to this where we allocate a `kj::Array<kj::byte>` with an attached memory adjustment that ends up outliving the isolate to ensure that the order of destruction is safe.